cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,158 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+
+    ref = vcs_url[-40:]  # Take the last 40 characters (the git ref)
+    clean_url = vcs_url[:-41]  # Drop the last 41 characters ('#' + git ref)
+
+    url = urlparse(clean_url)
+    namespace_repo = url.path.strip("/")
+    if namespace_repo.endswith(".git"):
+        namespace_repo = namespace_repo[: -len(".git")]
+
+    # Everything up to the last '/' is namespace, the rest is repo
+    namespace, _, repo = namespace_repo.rpartition("/")
+
+    vcs_url_info = {
+        "url": clean_url,
+        "ref": ref.lower(),
+        "host": str(url.hostname),
+        "namespace": namespace,
+        "repo": repo,
+    }
+
+    for key, value in vcs_url_info.items():
+        if not value:
+            raise FetchError(f"{vcs_url} is not valid VCS url! {key} is missing.")
+
+    return vcs_url_info
+
+
+def clone_repo_pack_archive(
+    vcs: str,
+    download_dir: RootedPath,
+) -> RootedPath:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param url: URL for file download
+    :param download_dir: Output folder where dependencies will be downloaded
+    :raise FetchError: If download failed
+    """
+    info = extract_git_info_npm(vcs)
+    download_path = RootedPath(
+        Path(
+            download_dir,
+            info["host"],  # host
+            info["namespace"],
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def get_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of dependencies with downloaded paths
+    """
+    files_to_download: dict[str, dict[str, Any]] = {}
+    download_paths = {}
+    for dependency, info in deps_to_download.items():
+        if info["url"].startswith("git+"):
+            # Git source
+            download_paths[dependency] = clone_repo_pack_archive(info["url"], download_dir)
+        else:
+            if "registry.npmjs.org" == urlparse(info["url"]).hostname:
+                # Tar archive from npm registry
+                archive_name = f'{dependency}-{info["version"]}.tgz'.removeprefix("@").replace(
+                    "/", "-"
+                )
+                download_paths[dependency] = download_dir.join_within_root(archive_name)
+            else:
+                # Tar archive
+                algorithm, digest = ChecksumInfo.from_sri(info["integrity"])
```

Comment:
I tried processing the cachito-npm-with-deps repository, and this function is causing an error when trying to get the checksum from [is-positive](https://github.com/cachito-testing/cachito-npm-with-deps/blob/master/package-lock.json#L52-L54).

It seems some git dependencies don't have the "git+" prefix.

Reply:
ou nice catch! @chmeliik is this a valid scenario as well? https://github.com/cachito-testing/cachito-npm-with-deps/blob/master/package-lock.json#L52-L54 when I was testing package-lock.json creation for each version 1-3 it walways created git+ssh version for VCS. I can add another check for both `git+` and `github:`

Reply:
Cachito recognizes these git prefixes: https://github.com/containerbuildsystem/cachito/blob/6b01c763018f7569d2aeabbd982e2381297fc7cb/cachito/workers/pkg_managers/general_js.py#L695

And uses `npm pack` to handle them all. If we don't want to use npm pack, we'll need special logic for `gitlab:`, `github:` and `bitbucket:`

Can you try re-creating the cachito-npm-with-deps lockfile with npm v6? If it uses a `git+<scheme>://` resolved url, maybe we don't care

Reply:
(the special logic probably wouldn't be that interesting, just change `{host}:{path}` to `git+ssh://{full_hostname}/{path}`

Reply:
I tested it with `npm:6.1.0` - http://pastebin.test.redhat.com/1101291 and package-lock.json were not updated with git+ssh. I'll add special logic.

Reply:
I believe this might help https://github.com/containerbuildsystem/cachi2/pull/152/files#diff-1bc123673d60c878542386b0a8ba8a27e381c4627075c4a5f3e106f0e74263acR202 I'll test it with unit test properly